File: scala/src/main/scala/HumanPlayer.scala
import scala.io.StdIn.readLine

class HumanPlayer(private val _name: String, private val _startingPosition: Int, private val _suggestions: Int) extends Player(_name, _startingPosition, _suggestions) {

  override def handleMoveSuggestion(initialRoll: Int): Int = {
    var finalRoll = initialRoll

    if (suggestions > 0) {
      val useSuggestion = readLine("You have " + suggestions + " remaining suggestions, do you want to use one? (y/n): ")

      if (useSuggestion == "y") {
        suggestions -= 1

        var newPos = getPosition + initialRoll

        if (GameBoard.isInSnakes(newPos)) {
          println("You would have hit a snake, rolling again...")

          while (GameBoard.isInSnakes(newPos)) {
            finalRoll = Die.roll()
            newPos = position + finalRoll
          }
        }
        else if (GameBoard.isInLadders(newPos)) {
          println("You will hit a ladder, moving player")
        }
        else {
          println("Safe to move, moving player...")
        }

        Thread.sleep(500)
      }
    }

    finalRoll
  }
}

File: C++/source/HumanPlayer.cc
//
// Created by liamg on 24/10/2025.
//

#include "../header/HumanPlayer.h"

#include <iostream>

#include <unistd.h>

#include "../header/GameBoard.h"

HumanPlayer::HumanPlayer(const string &name, const int position, const int suggestions) : Player(name, position, suggestions) {
}

int HumanPlayer::handleMoveSuggestion(const int initialRoll, Die* die) {
  int finalRoll = initialRoll;

  if (suggestions > 0) {
    cout << "You have " << suggestions
              << " remaining suggestions, do you want to use one? (y/n): ";

    char useSuggestion;
    cin >> useSuggestion;

    while (cin.get() != '\n') {}

    if (useSuggestion == 'y') {
      suggestions--;

      if (int newPos = position + initialRoll; GameBoard::isInSnakes(newPos)) {
        cout << "You would have hit a snake, rolling again..." << endl;

        while (GameBoard::isInSnakes(newPos)) {
          finalRoll = die->roll();
          newPos = position + finalRoll;
        }
      }
      else if (GameBoard::isInLadders(newPos)) {
        cout << "You will hit a ladder, moving player..." << endl;
      }
      else {
        cout << "Safe to move, moving player..." << endl;
      }
    }

    usleep(500000);
  }

  return finalRoll;
}